<a href="https://colab.research.google.com/github/pushkal1234/Automatic-Essay-Grading_NLP/blob/main/SHL_Pushkal_Shukla.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [210]:

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "/content/input/" directory
import os
for dirname, _, filenames in os.walk('/content/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

/content/input/all_prompts.csv
/content/input/train.csv
/content/input/test.csv


In [211]:
import os
import pandas as pd

df = pd.read_csv("/content/input/train.csv")
y = df['evaluator_rating']
df.drop(df.columns[[0]],axis=1, inplace= True)
df = df.dropna(axis=1)
df.head()

promptId  ... evaluator_rating
0         1  ...              3.0
1         1  ...              4.0
2         1  ...              2.0
3         1  ...              2.0
4         1  ...              3.0

[5 rows x 4 columns]

In [ ]:
df['essay'][1]

"I am agree the tightly defined curriculum of our education system leaves no room for imagination and creativity.A hidden curriculum can be defined as the lesson that are taught informally,and usually unintentionally in the school and colleges. These include behaviour ,perspectives and attitudes that students pick up while they are at school.This is contrasted with the formal curriculum,such as the courses and activities students participate in,The curriculum began early in the child's education. Students learn to form opinion and ideas about their environment and their classmate.They also learn what is expected of them.These attitudes and ideas are not taught in formal way,but kids absorb them through natural observation and participation in classroom more and more and also participate in social activities.Areas of hidden curriculum in our school that mold perspectives of students deal with issues such as gender,morals,social class,politics and language.Many books and magzine which ar

In [212]:
import warnings
warnings.filterwarnings("ignore")                     #Ignoring unnecessory warnings

import numpy as np                                  #for large and multi-dimensional arrays
import pandas as pd                                 #for data manipulation and analysis
import nltk                                         #Natural language processing tool-kit

from nltk.corpus import stopwords                   #Stopwords corpus
from nltk.stem import PorterStemmer                 # Stemmer

from sklearn.feature_extraction.text import CountVectorizer          #For Bag of words
from sklearn.feature_extraction.text import TfidfVectorizer          #For TF-IDF
from gensim.models import Word2Vec                                   #For Word2Vec

from tensorflow.keras.layers import Embedding
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.models import Sequential
from tensorflow.keras.preprocessing.text import one_hot
from tensorflow.keras.layers import LSTM
from tensorflow.keras.layers import Dropout
from tensorflow.keras.layers import Dense

In [ ]:
df.shape

(1240, 4)

In [215]:
list_1 = []
list_2 = []
list_3 = []
list_4 = []
list_5 = []
list_6 = []
list_7 = []
list_8 = []
list_9 = []
list_10 = []
list_11 = []

In [216]:
data = []
num = [0,0.5,1,1.5,2,2.5,3,3.5,4,4.5,5]
count = 0
for i,j in zip(range(len(df)),df["essay"]):
    if df["evaluator_rating"][i] == 0:
        list_1.append(j)
    if df["evaluator_rating"][i] == 0.5:
        list_2.append(j)
    if df["evaluator_rating"][i] == 1:
        list_3.append(j)
    if df["evaluator_rating"][i] == 1.5:
        list_4.append(j)
    if df["evaluator_rating"][i] == 2:
        list_5.append(j)
    if df["evaluator_rating"][i] == 2.5:
        list_6.append(j)
    if df["evaluator_rating"][i] == 3:
        list_7.append(j)
    if df["evaluator_rating"][i] == 3.5:
        list_8.append(j)
    if df["evaluator_rating"][i] == 4:
        list_9.append(j)
    if df["evaluator_rating"][i] == 4.5:
        list_10.append(j)
    if df["evaluator_rating"][i] == 5:
        list_11.append(j)
   

In [217]:
import itertools
data = list(itertools.chain(list_1,list_2,list_3,list_4,list_5,list_6,list_7,list_8,list_9,list_10,list_11))
len(data)

1240

In [192]:
def gen_num(num,length):
    num_list = [num]*length
    return num_list

In [220]:
score_1 = gen_num(0,len(list_1))
score_2 = gen_num(0.5,len(list_2))
score_3 = gen_num(1,len(list_3))
score_4 = gen_num(1.5,len(list_4))
score_5 = gen_num(2,len(list_5))
score_6 = gen_num(2.5,len(list_6))
score_7 = gen_num(3,len(list_7))
score_8 = gen_num(3.5,len(list_8))
score_9 = gen_num(4,len(list_9))
score_10 = gen_num(4.5,len(list_10))
score_11 = gen_num(5,len(list_11))

score = list(itertools.chain(score_1,score_2,score_3,score_4,score_5,score_6,score_7,score_8,score_9,score_10,score_11))
len(score)


1240

In [221]:
# dictionary of lists 
dictnary = {'essay': data, 'score': score}     
df = pd.DataFrame(dictnary) 
df 

essay  score
0     Education is the most important part of our li...    0.0
1     Our education inindia was good,but not that mu...    0.0
2     Education is very important in each and every ...    0.0
3     education system is the most important for us ...    0.0
4     today's eduction system is very bad due to res...    0.0
...                                                 ...    ...
1235  Imagination is the mother of invention and wit...    5.0
1236  Just learning a text-book by heart and getting...    5.0
1237  YES I AGREE WITH THIS STATEMENT THAT THE TIGHT...    5.0
1238  I speak about education from an unflattering p...    5.0
1239  Yes, it is true that the gap between rich and ...    5.0

[1240 rows x 2 columns]

In [222]:
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [223]:
stop_words = set(stopwords.words('english'))
len(stop_words) #finding stop words

179

In [224]:
import re
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer
snow = nltk.stem.SnowballStemmer('english')

corpus = []
for i in range(0, len(df)):
    review = re.sub('[^a-zA-Z]', ' ', df['essay'][i])
    review = review.lower()
    review = review.split()
    review = [snow.stem(word) for word in review if not word in stopwords.words('english')]
    review = ' '.join(review)
    corpus.append(review)

In [225]:
voc_size=5000
onehot_repr=[one_hot(words,voc_size)for words in corpus] 
type(onehot_repr)

list

In [226]:
sent_length=500
embedded_docs=pad_sequences(onehot_repr,padding='pre',maxlen=sent_length)
print(embedded_docs)

[[   0    0    0 ... 3456  844 2884]
 [   0    0    0 ... 3846 1241 3255]
 [   0    0    0 ... 4853 1505 1888]
 ...
 [   0    0    0 ... 4912 1729 1505]
 [   0    0    0 ... 2530 1549 2686]
 [   0    0    0 ... 3125 1991 3218]]


In [ ]:
df.shape

(1240, 2)

Feature Extraction Functions and conversion of them into Vector form so that we could feed our model into deep learning module

In [227]:
import numpy as np
import nltk
import re
from nltk.corpus import stopwords
from gensim.models import Word2Vec

def essay_to_wordlist(essay_v, remove_stopwords):
    """Remove the tagged labels and word tokenize the sentence."""
    essay_v = re.sub("[^a-zA-Z]", " ", essay_v)
    words = essay_v.lower().split()
    if remove_stopwords:
        stops = set(stopwords.words("english"))
        words = [w for w in words if not w in stops]
    return (words)

def essay_to_sentences(essay_v, remove_stopwords):
    """Sentence tokenize the essay and call essay_to_wordlist() for word tokenization."""
    tokenizer = nltk.data.load('tokenizers/punkt/english.pickle')
    raw_sentences = tokenizer.tokenize(essay_v.strip())
    sentences = []
    for raw_sentence in raw_sentences:
        if len(raw_sentence) > 0:
            sentences.append(essay_to_wordlist(raw_sentence, remove_stopwords))
    return sentences

def makeFeatureVec(words, model, num_features):
    """Make Feature Vector from the words list of an Essay."""
    featureVec = np.zeros((num_features,),dtype="float32")
    num_words = 0.
    index2word_set = set(model.wv.index2word)
    for word in words:
        if word in index2word_set:
            num_words += 1
            featureVec = np.add(featureVec,model[word])        
    featureVec = np.divide(featureVec,num_words)
    return featureVec

def getAvgFeatureVecs(essays, model, num_features):
    """Main function to generate the word vectors for word2vec model."""
    counter = 0
    essayFeatureVecs = np.zeros((len(essays),num_features),dtype="float32")
    for essay in essays:
        essayFeatureVecs[counter] = makeFeatureVec(essay, model, num_features)
        counter = counter + 1
    return essayFeatureVecs

In [228]:
import nltk

nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [229]:
from keras.layers import Embedding, LSTM, Dense, Dropout, Lambda, Flatten
from keras.models import Sequential, load_model, model_from_config
import keras.backend as K

def get_model():
    """Define the model."""
    model = Sequential()
    model.add(LSTM(300, dropout=0.4, recurrent_dropout=0.4, input_shape=[1, 300], return_sequences=True))
    model.add(LSTM(64, recurrent_dropout=0.4))
    model.add(Dropout(0.5))
    model.add(Dense(1, activation='relu'))

    model.compile(loss='mean_squared_error', optimizer='rmsprop', metrics=['accuracy','mae'])
    model.summary()

    return model

In [230]:
y = df['score']

In [205]:
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [ ]:
df.shape

(1240, 2)

In [207]:
from sklearn.model_selection import KFold
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from sklearn.metrics import max_error
cv = KFold(n_splits = 10, random_state=1, shuffle = True)
results = []
y_pred_list = []

count = 1
for traincv, testcv in cv.split(df):
    print("\n--------Fold {}--------\n".format(count))
    df_test, df_train, y_test, y_train = df.iloc[testcv], df.iloc[traincv], y.iloc[testcv], y.iloc[traincv]
    
    train_essays = df_train['essay']
    test_essays = df_test['essay']
    
    sentences = []
    
    for essay in train_essays:
            # Obtaining all sentences from the training essays.
            sentences += essay_to_sentences(essay, remove_stopwords = True)
            
    # Initializing variables for word2vec model.
    num_features = 300 
    min_word_count = 40
    num_workers = 4
    context = 10
    downsampling = 1e-3

    print("Training Word2Vec Model...")
    model = Word2Vec(sentences, workers=num_workers, size=num_features, min_count = min_word_count, window = context, sample = downsampling)

    model.init_sims(replace=True)
    model.wv.save_word2vec_format('word2vecmodel.bin', binary=True)

    clean_train_essays = []
    
    # Generate training and testing data word vectors.
    for essay_v in train_essays:
        clean_train_essays.append(essay_to_wordlist(essay_v, remove_stopwords=True))
    trainDataVecs = getAvgFeatureVecs(clean_train_essays, model, num_features)
    
    clean_test_essays = []
    for essay_v in test_essays:
        clean_test_essays.append(essay_to_wordlist( essay_v, remove_stopwords=True ))
    testDataVecs = getAvgFeatureVecs( clean_test_essays, model, num_features )
    
    trainDataVecs = np.array(trainDataVecs)
    testDataVecs = np.array(testDataVecs)
    # Reshaping train and test vectors to 3 dimensions. (1 represnts one timestep)
    trainDataVecs = np.reshape(trainDataVecs, (trainDataVecs.shape[0], 1, trainDataVecs.shape[1]))
    testDataVecs = np.reshape(testDataVecs, (testDataVecs.shape[0], 1, testDataVecs.shape[1]))
    
    lstm_model = get_model()
    lstm_model.fit(trainDataVecs, y_train, batch_size=64, epochs= 50)
    #lstm_model.load_weights('./model_weights/final_lstm.h5')
    y_pred = lstm_model.predict(testDataVecs)
    
    # Save any one of the 5 models.
    if count == 5:
         lstm_model.save('./final_lstm.h5')
    
    # Round y_pred to the nearest integer.
    y_pred = np.around(y_pred)
    
    # Evaluate the model on the evaluation metric. "Since This is an Regression Problem we will be using R2_Score"
    from sklearn.metrics import max_error
    result = r2_score(y_test.values, y_pred, multioutput='variance_weighted')
    acc = max_error(y_test.values,y_pred)
    print("acc Score: {}".format(acc))
    print("R2 Score: {}".format(result))
    results.append(result)

    count += 1



--------Fold 1--------

Training Word2Vec Model...
Model: "sequential_18"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
lstm_36 (LSTM)               (None, 1, 300)            721200    
_________________________________________________________________
lstm_37 (LSTM)               (None, 64)                93440     
_________________________________________________________________
dropout_18 (Dropout)         (None, 64)                0         
_________________________________________________________________
dense_18 (Dense)             (None, 1)                 65        
Total params: 814,705
Trainable params: 814,705
Non-trainable params: 0
_________________________________________________________________
Epoch 1/50
146/146 [==============================] - 13s 30ms/step - loss: 7.6686 - accuracy: 0.1219 - mae: 1.9966
Epoch 2/50
146/146 [==============================] - 4s 30ms/step - loss: 2.

In [209]:
print("Average R2 score after a 10-fold cross validation: ",np.around(np.array(results).mean(),decimals=4))

Average R2 score after a 10-fold cross validation:  0.8657


Hence we can find Above after running 10 Fold Cross Validation we get R2 Score as 0.8657 
This model Performs well in such Small Dataset as well.


In [ ]:
#content = pd.read_csv("/content/input/test.csv")
testset = pd.read_csv("/content/input/test.csv")
testset.drop(testset.columns[[0]],axis=1, inplace= True)
testset = testset.dropna(axis=1)
testset

promptId uniqueId                                              essay
0           1    1_315  Curriculum has been adopted in many schools. T...
1           1    1_214  I strongly agree with the statement ,  The tig...
2           1    1_196  Imagination and creativity is the most importa...
3           1    1_178  In our eduction system leaves no room for imag...
4           1    1_201  I will agree at some what extend, because if w...
..        ...      ...                                                ...
300         5    5_146  Earth is a creation of God and everything that...
301         5     5_65  production of arms and weapons in this present...
302         5    5_151  Race to become more powerful can destroy the e...
303         5    5_404  In its attempt to harness the power of the ato...
304         5    5_360  Racein the production of arms and weapons in t...

[305 rows x 3 columns]

Now, Moving towards subimission.csv file extraction from the weights as learned in the module, we will apply same weights and bias to get the prediction

In [ ]:
x1 = testset['essay'][1]
x1

"I strongly agree with the statement ,  The tightly defined curriculum of our education system leaves no room or imagination and creativityToday's education system is completely theory based. More focus is given on theory,text books and pre-defined things set. No room for practical things,experiments, innovation and creativity.Our education system is mainly rather completely pre-defined curriculum structure as to syllabus is concerned , then examination pattern, marking system is all decided by the respective state government. In each state curriculum followed is different and preset by our older generation/people. No new concepts are added , no innovative systems are created.Syallbus is changed once in every ten years, which is a long gap and in between various things are changing , the world around is changing rapidly but our education system is still lacking behind. Also our system in based on gradation which creates inferority among them.If you see in western part of the world, thi

In [ ]:
x2 = testset['essay'][2]
x2

'Imagination and creativity is the most important thing to develope any idea and innovation.if you have no of us than is to difficults to innovation.In education we have no room for that things than our education is menningless.because when we do imagination after that we create somethings.so imagination and creativity is value able in our room so nothing happen in our education. when we saw any thing immediatly we starting the imagination and than we make something.so its important things in our life.without this we cannot make or create any things.so imagination and creativity is most important think nowdays. if it has no seperate room for imagination and creativity so out that any peple thing or imagin wrong thing or create that thing so its to harmfull to us.we take one example on this currently one of most concerning aspects of raising children is the amount of time they spend in front of screens,whether it tv,video etc,imagination and creativity is most important thing to develop

In [ ]:
x3 = testset['essay'][3]
x3

'In our eduction system leaves no room for imagination and creativity.In education system some staff are applied the rules and regulation.The staff are forcing the do many things at a time.Student brain are very small they cannot do many thing at a time.The staff are gives the many activity and said to student complete one or two days but this activity is big then student not completed at given time its need to complete given task at more extra time.Now a days the staff are not teach perfectly to the particular subject and not give the explanation of topics its just writing topic name and give the two or three line explanation.Now a days the some students going to school,college regularly and attend the all lectures but she/he cannot getting on what actually teach and how its work when and who is developed.In this situation and case the some students are feel tightly.In eduction system do not applied force for any students.The subject is teach very carefully and topic is explained by p

In [ ]:
x5 = testset['essay'][4]
x5

'I will agree at some what extend, because if we think on regular basic of university of mumbai curriculam we didnt get time to do different things simelteneously with the study, study having lot of assignments, sheets, practicle even a theory,for that we have to come in the clg for 8 hours in travelling 2 hours wasted every day,9 hours went in the sleep,5 hour remaining in which we do our daily work, over an all the shedule is tight and we are not able to do creative things and for creation imagination is needed. this is one side of our lifeBut on other hand if we thought like in clg we have many practicle,events and also free time , in that free time regarding to the curriculum taking it as creation and make it more innovative and interesting as an project that will be awsome , it is tough to do but not an impossible , every work is hard but not imppossible. we can change in our daily shedule to give time for creativity and imagination. that would be more preferable and good for heal

We are having 4 Sample essay as provided in our Dataset, Now we will run our predictive Analysis to get score of these eassy and extract out Test_submission.csv file for further Evaluation.
 

In [ ]:
from gensim.models import Word2Vec
from gensim.models import KeyedVectors
num_features = 300
      
model = KeyedVectors.load_word2vec_format( "/content/word2vecmodel.bin", binary=True)
clean_test_essays = []
clean_test_essays.append(essay_to_wordlist(x1, remove_stopwords=True ))
testDataVecs = getAvgFeatureVecs( clean_test_essays, model, num_features )
testDataVecs = np.array(testDataVecs)
testDataVecs = np.reshape(testDataVecs, (testDataVecs.shape[0], 1, testDataVecs.shape[1]))

# lstm_model = get_model()
lstm_model.load_weights("/content/final_lstm.h5")
preds = lstm_model.predict(testDataVecs)

In [ ]:
pred1 = int(np.around(preds))
pred1

3

In [ ]:
pred2 = int(np.around(preds))
pred2

3

In [ ]:
pred3 = int(np.around(preds))
pred3

3

In [ ]:
pred5 = int(np.around(preds))
pred5

3

In [ ]:
testset.T.shape

(3, 305)

In [ ]:
y_pred1 = y_pred.T.reshape(206,)

In [ ]:
y_pred1

array([3., 2., 3., 3., 3., 2., 2., 2., 2., 2., 3., 3., 3., 2., 3., 2., 3.,
       3., 3., 3., 3., 2., 3., 2., 3., 2., 2., 2., 3., 3., 3., 3., 3., 3.,
       3., 3., 3., 2., 2., 3., 2., 3., 3., 2., 3., 2., 3., 2., 2., 3., 3.,
       3., 3., 3., 3., 3., 3., 3., 3., 3., 2., 2., 3., 2., 3., 3., 3., 3.,
       3., 3., 3., 3., 3., 3., 3., 2., 2., 2., 3., 2., 2., 2., 3., 2., 2.,
       3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 2., 2., 2., 3.,
       3., 2., 2., 3., 2., 3., 3., 2., 2., 3., 3., 3., 3., 3., 3., 3., 3.,
       3., 3., 3., 3., 3., 3., 3., 3., 3., 2., 2., 2., 2., 2., 3., 3., 3.,
       3., 3., 3., 3., 2., 2., 2., 3., 3., 3., 3., 2., 2., 3., 3., 3., 3.,
       3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3.,
       3., 3., 3., 3., 2., 3., 2., 2., 2., 2., 2., 2., 3., 3., 3., 3., 3.,
       3., 2., 3., 3., 3., 3., 3., 3., 3., 2., 2., 2., 3., 3., 3., 3., 3.,
       3., 3.], dtype=float32)

In [ ]:
outputfile = pd.DataFrame(y_pred1).to_csv("submission_Test.csv")

Done!!
It was an Excellent Problem Statement Involving multiple Concepts, thanks a lot to the team Involved. 